File: src/cachibot/models/bot.py
"""
Pydantic models for Bot configuration.

Bot data synced from frontend to enable platform message processing.
"""

from datetime import datetime

from pydantic import BaseModel, Field


class Bot(BaseModel):
    """Bot configuration stored in backend."""

    id: str
    name: str
    description: str | None = None
    icon: str | None = None
    color: str | None = None
    model: str
    system_prompt: str = Field(alias="systemPrompt")
    capabilities: dict = Field(default_factory=dict)
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    class Config:
        populate_by_name = True


class BotCreate(BaseModel):
    """Request body for creating/updating a bot."""

    id: str
    name: str
    description: str | None = None
    icon: str | None = None
    color: str | None = None
    model: str
    system_prompt: str = Field(alias="systemPrompt")
    capabilities: dict = Field(default_factory=dict)
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    class Config:
        populate_by_name = True


class BotResponse(BaseModel):
    """Response model for a bot."""

    id: str
    name: str
    description: str | None
    icon: str | None
    color: str | None
    model: str
    systemPrompt: str
    capabilities: dict
    createdAt: str
    updatedAt: str

    @classmethod
    def from_bot(cls, bot: Bot) -> "BotResponse":
        return cls(
            id=bot.id,
            name=bot.name,
            description=bot.description,
            icon=bot.icon,
            color=bot.color,
            model=bot.model,
            systemPrompt=bot.system_prompt,
            capabilities=bot.capabilities,
            createdAt=bot.created_at.isoformat(),
            updatedAt=bot.updated_at.isoformat(),
        )

File: mobile/lib/models/bot.dart
class Bot {
  const Bot({
    required this.id,
    required this.name,
    this.description,
    this.icon,
    this.color,
    required this.model,
    this.models,
    required this.systemPrompt,
    required this.capabilities,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String name;
  final String? description;
  final String? icon;
  final String? color;
  final String model;
  final Map<String, dynamic>? models;
  final String systemPrompt;
  final Map<String, dynamic> capabilities;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory Bot.fromJson(Map<String, dynamic> json) {
    return Bot(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      icon: json['icon'] as String?,
      color: json['color'] as String?,
      model: json['model'] as String,
      models: json['models'] as Map<String, dynamic>?,
      systemPrompt: json['systemPrompt'] as String,
      capabilities: (json['capabilities'] as Map<String, dynamic>?) ?? {},
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'description': description,
        'icon': icon,
        'color': color,
        'model': model,
        'models': models,
        'systemPrompt': systemPrompt,
        'capabilities': capabilities,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };
}
